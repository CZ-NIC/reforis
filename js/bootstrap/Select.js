/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import {LABEL_SIZE, FIELD_SIZE} from "./constants";


export default function Select({name, id, label, choices, value, onChange, ...props}) {
    const options = choices.map((choice, key) => {
        return <option key={key} value={choice.value}>{choice.label}</option>;
    });

    return <div className="form-group row">
        <label className={"form-control-label col-sm-" + LABEL_SIZE} htmlFor={id}>{label}</label>
        <div className={"col-sm-" + FIELD_SIZE}>
            <select
                className="form-control"
                id={id}
                name={name}
                value={value}
                onChange={onChange}
                {...props}
            >
                {options}
            </select>
        </div>
    </div>;
}


