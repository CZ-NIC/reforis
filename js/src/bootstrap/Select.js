/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import {LABEL_SIZE, FIELD_SIZE} from './constants';


export default function Select({name, id, label, choices, value, onChange, disabled, helpText, ...props}) {

    let options = [];
    for (let key in choices)
        if (choices.hasOwnProperty(key))
            options.push(<option key={key} value={key}>{choices[key]}</option>);

    return <div className='form-group row'>
        <label className={'form-control-label col-sm-' + LABEL_SIZE} htmlFor={id}>{label}</label>
        <div className={'col-sm-' + FIELD_SIZE}>
            <select
                disabled={disabled}
                className='form-control'
                id={id}
                name={name}
                value={value}
                onChange={onChange}
                {...props}
            >
                {options}
            </select>
            <small className="form-text text-muted">{helpText}</small>
        </div>
    </div>;
}


