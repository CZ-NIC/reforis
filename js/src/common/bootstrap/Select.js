/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import PropTypes from 'prop-types';
import {useUID} from 'react-uid/dist/es5/index';


Select.propTypes = {
    /** Select field Label. */
    label: PropTypes.string.isRequired,
    /** Choices if form of {value : "Label",...}.*/
    choices: PropTypes.object.isRequired,
    /** Current value. */
    value: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
    ]).isRequired,
    /** Help text message. */
    helpText: PropTypes.string,
};

export default function Select({label, choices, helpText, ...props}) {
    const uid = useUID();

    const options = Object.keys(choices).map(
        key => <option key={key} value={key}>{choices[key]}</option>
    );

    return <div className='form-group col-sm-12 offset-lg-1 col-lg-10'>
        <label htmlFor={uid}>{label}</label>
        <select
            className='custom-select'
            id={uid}
            {...props}
        >
            {options}
        </select>
        {helpText ? <small className="form-text text-muted">{helpText}</small> : null}
    </div>;
}
