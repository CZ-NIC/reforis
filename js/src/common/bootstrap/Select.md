Select with options Bootstrap component input with label and predefined sizes and structure for using in foris forms.

All additional `props` are passed to the `<select>` HTML component.

```js
import {useState} from 'react';
const CHOICES={
    apple:'Apple',
    banana:'Banana',
    peach:'Peach',
};
const [value, setValue] = useState(Object.keys(CHOICES)[0]);

<>
    {/*Yeah, it gets event, not value!*/}
    <Select
        label="Fruit"
        value={value}
        choices={CHOICES}
        onChange={event=>setValue(event.target.value)}
    />
    <p>Selected choice label: {CHOICES[value]}</p>
    <p>Selected choice value: {value}</p>
</>
```
