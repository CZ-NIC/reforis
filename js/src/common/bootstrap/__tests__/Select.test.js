/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import {render} from 'react-testing-library';

import Select from '../Select';

const TEST_CHOICES = {
    key1: 'value1',
    key2: 'value2',
    key3: 'value3',
};

describe('<Select/>', () => {
    it('Render select', () => {
        const {container} = render(
            <Select
                label='Test label'
                value='value'
                choices={TEST_CHOICES}
                helpText={'Help text'}
                onChange={() => {
                }}
            />
        );
        expect(container.firstChild).toMatchSnapshot();
    });
});