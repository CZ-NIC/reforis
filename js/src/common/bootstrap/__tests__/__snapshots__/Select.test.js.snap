// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select/> Test with snapshot. 1`] = `
<div>
  <div
    class="form-group col-sm-12 offset-lg-1 col-lg-10"
  >
    <label
      for="1"
    >
      Test label
    </label>
    <select
      class="custom-select"
      id="1"
    >
      <option
        value="1"
      >
        one
      </option>
      <option
        value="2"
      >
        two
      </option>
      <option
        value="3"
      >
        three
      </option>
    </select>
    <small
      class="form-text text-muted"
    >
      Help text
    </small>
  </div>
</div>
`;
